public/app/features/integrated-alerting/components/Table/Pagination/Pagination.tsx

```diff
@@ -0,0 +1,125 @@
+import React, { FC, useState, useMemo, useEffect } from 'react';
+import { useStyles, IconName, Button, Select } from '@grafana/ui';
+import { PaginationProps } from './Pagination.types';
+import { getStyles } from './Pagination.styles';
+import { Messages } from './Pagination.messages';
+
+export const Pagination: FC<PaginationProps> = ({
+  pageCount: propPageCount,
+  initialPageIndex,
+  pageSize,
+  pagesPerView,
+  nrRowsOnCurrentPage,
+  totalItems,
+  pageSizeOptions,
+  onPageChange,
+  onPageSizeChange,
+}) => {
+  // Zero pages probably won't make the pagination show up, but here we should be agnostic to that
+  const [pageCount, setPageCount] = useState(Math.max(propPageCount, 1));
+  const [activePageIndex, setActivePageIndex] = useState(initialPageIndex);
+  const pageArray = useMemo(() => [...Array(pageCount).keys()], [pageCount]);
+  const maxVisiblePages = Math.min(pagesPerView, pageCount);
+  // We want to center our selected page, thus we need to know how many should be on the left
+  const pagesBehind = pagesPerView - (pagesPerView - Math.ceil(pagesPerView / 2)) - 1;
+  let firstPageIndex = Math.max(activePageIndex - pagesBehind, 0);
+  let lastPageIndex = firstPageIndex + maxVisiblePages;
+
+  // If we can't keep the selected page in the center anymore, it should just move rightwards
+  if (lastPageIndex >= pageCount + 1 && lastPageIndex - maxVisiblePages > 0) {
+    lastPageIndex = pageCount;
+    firstPageIndex = lastPageIndex - maxVisiblePages;
+  }
+  const shownPages = pageArray.slice(firstPageIndex, lastPageIndex);
+  const leftItemNumber = propPageCount > 0 ? activePageIndex * pageSize + 1 : 0;
+  const rightItemNumber = activePageIndex * pageSize + nrRowsOnCurrentPage;
```

Comment:
I think this logic (or at least part of it) should be refactored and brought outside. Maybe using a hook?

Reply:
Done. Moved these into utils, I think a hook would be too much for something that won't be reused in other parts